[2주차] Counter 앱 만들고 파일 분리하기

hanna244 opened the pull request:
안녕하세요! 

작업을 진행하기 전 지난 과제의 여운이 남았는지 useState를 사용하지 않고 구현해야한다는 '이상한'생각에 초반에 시간을 많이 뺏겼습니다! 

아래와 같은 고민으로 컴포넌트를 구분하였습니다. 
- 결과값이 출력되는 컴포넌트 
- 그 이외 컨텐츠를 보며 성격이 비슷한 것끼리 묶기

코드리뷰 부탁드립니다 감사합니다 :-)

hanna244 committed:
update: count 구현

hanna244 committed:
update: 컴포넌트 파일 분리

hanna244 committed:
update: eslint 대응
- must be placed on a new line

hanna244 commented:
질문
1. eslint에서 컴포넌트를 export로 내보내지 말고 export default로 내보내도록 권장하는데요 export보다 export default를 권장하는 이유가 무엇인가요?

hanna244 commented on src/RenderButton.jsx:
```diff
@@ -0,0 +1,10 @@
+    <button type="button" onClick={() => handleClick(updateValue)} value={count}>
+      Click me!
+      {`(${count})`}
+    </button>
```
RenderButton 함수에 "Click me!" text를 넣을지 / prop로 text를 전달할지 잠시 고민했습니다.

결국 RenderButton 함수에 "Click me!" text를 넣는 것으로 결정한 이유는 현재 구현헤야하는 시안에서 해당 컴포넌트는 2번 이상 재사용되지 않기 때문입니다.

리뷰어님께서는 컴포넌트의 수정 가능 여부 또는 재사용성을 대비하여 대부분의 컨텐츠를 props로 변경할 수 있도록 하는 것을 어떻게 생각하시는지 궁금합니다!

KJoobin replied:
더 복잡한 컴포넌트에서는 재사용을 대비해 작성하는편이지만, 
지금 처럼 간단한 경우에서는 당장 컴포넌트를 만들때 필요한 props 만 전달합니다! 추후에 새로운 기능, 재사용을 위해 변경되는 부분이 있다면 그때가서 수정을 하면 되거든요!
소프트웨어는 하드웨어 랑은 다르게 변경이 쉽답니다 :)

hanna244 replied:
그렇군요! 처음부터 많은 props를 미리 작성하지 않아도 추후 필요할 때 수정하면 되겠네요 답변감사합니다!

hanna244 committed:
refactoring:  NumberButton 컴포넌트 더 작은 단위로 변경
- 컴포넌트 내부에서 map이 돌지 않도록 수정

KJoobin commented on src/NumberButton.jsx:
```diff
@@ -0,0 +1,9 @@
+
+export default function NumberButton({ handleClick, value }) {
+  return (
+    <button type="button" value={value} onClick={(e) => handleClick(e)}>
```
```suggestion
    <button type="button" value={value} onClick={handleClick}>
```
이렇게 작성하는게 가독성 측면에서 좋을거같은데 어떻게 생각하시나요!?

hanna244 replied:
넵! event 파라미터를 전달 한다는 것을 명시적으로 넣으면 어떨까 하는 생각으로 작성했습니다. 하지만 추후 가독성을 위해 말씀하신 것처럼  작성하는 것이 제가 봐도 좋은 것 같습니다!

KJoobin commented on src/index.jsx:
```diff
@@ -0,0 +1,41 @@
+import NumberButton from './NumberButton';
+
+function App() {
+  const [render, setRender] = useState(0);
```
render 보다 더 좋은 변수명이 있을까요?!

hanna244 replied:
더한 값의 결과라는 의미로 변수 네이밍 변경했습니다! 
render  →  sumResultValue

KJoobin commented on src/index.jsx:
```diff
@@ -0,0 +1,41 @@
+
+  /** 숫자 버튼 클릭 시 해당 값을 가져와 값을 합하는 이벤트 */
+  const handleGetUpdateValue = (e) => {
+    const value = Number(e.target.value);
```
`const value = Number(e.target.value)` target value 를 가져와서 Number 로 타입변환 하는걸 NumberButton 안에서 하는건 어떨까요?!

hanna244 replied:
코드를 수정하면서 Number()가 필요하지 않아서 삭제했습니다. 
button의 value 속성 값을 가져오는 것이 아닌 NumberButton 컴포넌트가 전달 받는 index 값을 바로 handleGetUpdateValue 함수의 인자로 전달 하는 것으로 수정했기 때문입니다. 

질문 : e.target.value의 속성 값의 타입은 왜 string인가요? NumberButton 컴포넌트에 전달 한 number를 <button value={value}> 속성에 그대로 적용했는데 왜 값의 타입이 변경되는 건가요? 
(* NumberButton이 전달 받는 value prop의 값의 type은 number입니다)

KJoobin replied:
한번도 고민해보지 않았던 질문이네요..! 아마 button 의 value 의 type 이 string 이기 떄문 아닐까요?!
[MDN button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/button)

KJoobin commented on src/index.jsx:
```diff
@@ -0,0 +1,41 @@
+  return (
+    <div>
+      <p>Counter</p>
+      <RenderButton count={render} updateValue={1} handleClick={handleAddValue} />
```
RenderButton 이라는 이름도 어색한거같아요!! 좀더 직관적으로 작성하면 좋을거같습니다 :)

hanna244 replied:
MainButton으로 네이밍 변경 진행했습니다!

KJoobin commented on src/index.jsx:
```diff
@@ -0,0 +1,41 @@
+  );
+}
+
+ReactDOM.render(<App />, document.getElementById('app'));
```
index.js 파일에서 App.jsx 파일을 분리해서 작성해볼까요?! (React 에서 index.js 의 역할, App.jsx 의 역할 에 대해서 검색해보면 좋을거같아요! )

KJoobin reviewed (commented):
과제 1번을 잘 진행해주신거같아요 !! 

네이밍만 조금 신경써서 작업을 해주시면 완벽할거같습니다 :)

hanna244 committed:
update: 1차 코드리뷰 확인 및 적용
- 변수명 수정
- NumberButton의 value 값에 Number() 사용하는 위치 변경
- index.js/ App.js 파일 분할
- 사용하지 않는 value 속성 삭제

hanna244 committed:
update: Number() 제거

KJoobin commented:
> eslint에서 컴포넌트를 export로 내보내지 말고 export default로 내보내도록 권장하는데요 export보다 export default를 권장하는 이유가 무엇인가요?

싱글 export 일 경우에는 export default 를 사용하라는 뜻 입니다! :)
[import/prefer-default-export](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/prefer-default-export.md)

KJoobin commented on src/App.jsx:
```diff
@@ -0,0 +1,39 @@
+import NumberButton from './NumberButton';
+
+function App() {
+  const [sumResultValue, setSumResultValue] = useState(0);
```
간단하게 `count` 는 어떠세요?! sumResultValue 의 경우에는 state 라고 보기 힘들거같고, result 와 value 가 중복되는 느낌이 있습니다 :)

KJoobin commented on src/App.jsx:
```diff
@@ -0,0 +1,39 @@
+  };
+
+  /** 숫자 버튼 클릭 시 해당 값을 가져와 값을 합하는 이벤트 */
+  const handleGetUpdateValue = (currentValue) => {
+    handleAddValue(currentValue);
+  };
```
보통 Get 은 정보를 가지고 오는 함수에서 많이 사용해요!!

KJoobin commented on src/MainButton.jsx:
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
```
react 17 부터는 React 를 Import 하지 않을수있는데 알고계신가요~ :)

KJoobin reviewed (commented):
과제를 잘 진행해주셨습니다! :)

data, value, result 와 같은 네이밍은 추상적이기 때문에 최대한 피하는게 좋아요!!

주석을 쓰지 않더라도 이해할수 있을정도의 네이밍을 고민하시면 좋을거같습니다 :)

KJoobin commented:
회고를 작성해주시면 머지하겠습니다!
